Bin/Snapshot.php

```diff
@@ -273,16 +304,53 @@ function () use ($tags, $newTag, $repositoryRoot, &$changelog) {
                             $title = $newTag;
                         }
 
-                        $changelog .=
-                            '# ' . $title . "\n\n" .
-                            Console\Processus::execute(
-                                'git --git-dir=' . $repositoryRoot . '/.git ' .
-                                    'log ' .
-                                        '--first-parent ' .
-                                        '--pretty="format:  * %s (%aN, %aI)" ' .
-                                        $fromStep . '...' . $toStep,
-                                false
-                            ) . "\n\n";
+                        $changelog .= '# ' . $title . "\n\n";
+
+                        $commits = Console\Processus::execute(
+                            'git --git-dir=' . $repositoryRoot . '/.git ' .
+                                'log ' .
+                                    '--first-parent ' .
+                                    '--pretty="format:  * %s (%aN, %aI)" ' .
+                                    $fromStep . '...' . $toStep,
+                            false
+                        );
+
+                        $collection = explode(PHP_EOL, $commits);
+                        $pretty = [];
+                        $otherCommits = [];
+                        foreach ($collection as $line) {
+                            if (preg_match($changelogThemesRegex, $line, $matches)) {
+                                if (!isset($pretty[$matches[1]])) {
+                                    $pretty[$matches[1]] = [$matches[2] => []];
+                                } elseif (!isset($pretty[$matches[1]][$matches[2]])) {
+                                    $pretty[$matches[1]][$matches[2]] = [];
+                                }
+
+                                $pretty[$matches[1]][$matches[2]][] = ' *' . $matches[3];
+                            } elseif (false === $onlyStructuredCommits) {
+                                $otherCommits[] = $line;
+                            }
+                        }
+
+                        foreach ($pretty as $theme => $contexts) {
+                            $userFriendlyTheme = self::CHANGELOG_THEMES[$theme] ?? $theme;
+                            $changelog .= '## ' . $userFriendlyTheme . "\n\n";
+
+                            foreach ($contexts as $name => $context) {
+                                $userFriendlyName = strlen($name) < 4
+                                    ? strtoupper($name)
+                                    : ucfirst($name);
+                                $changelog .=
+                                    '### ' . $userFriendlyName . "\n\n" .
+                                    implode(PHP_EOL, $context) . "\n\n";
```

Comment:
Either you use `PHP_EOL` or `"\n"`, but please don't use both.

My personal take: `"\n"`.